internal/witness/cmd/feeder/main.go

```diff
@@ -23,140 +23,175 @@ import (
+		return nil, fmt.Errorf("failed to read HTTP response: %v", err)
 	}
-	wv, err := verifier.NewWitnessVerifier(pk)
+	// Get data for all usable logs.
+	logList, err := loglist2.NewFromJSON(body)
 	if err != nil {
-		glog.Exitf("Failed to create witness signature verifier: %v", err)
+		return nil, fmt.Errorf("failed to parse JSON: %v", err)
+	}
+	logs := make(map[string]logData)
+	usable := logList.SelectByStatus([]loglist2.LogStatus{loglist2.UsableLogStatus})
+	for _, operator := range usable.Operators {
+		for _, log := range operator.Logs {
+			logID := base64.StdEncoding.EncodeToString(log.LogID)
+			c, err := createLogClient(log.Key, log.URL)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create log client: %v", err)
+			}
+			logs[logID] = logData{
+				desc:   log.Description,
+				client: c,
+			}
+		}
 	}
+	return logs, nil
+}
+
+// createLogClient creates a CT log client from a public key and URL.
+func createLogClient(key []byte, url string) (*client.LogClient, error) {
+	pemPK := pem.EncodeToMemory(&pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: key,
+	})
+	opts := jsonclient.Options{PublicKey: string(pemPK)}
+	c, err := client.New(url, http.DefaultClient, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create JSON client: %v", err)
+	}
+	return c, nil
+}
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+	// Set up the witness client.
+	var w wh.Witness
 	if wURL, err := url.Parse(*witness); err != nil {
 		glog.Exitf("Failed to parse witness URL: %v", err)
 	} else {
 		w = wh.Witness{
-			URL:      wURL,
-			Verifier: *wv,
+			URL: wURL,
 		}
 	}
-	// Now set up the log client.
-	logID, err := wa.LogIDFromPubKey(*logPK)
+	// Now set up the log data (with no initial witness STH).
+	logData, err := populateLogs(*logList)
 	if err != nil {
-		glog.Exitf("Failed to create log id: %v", err)
+		glog.Exitf("Failed to set up log data: %v", err)
 	}
-	if *logURL == "" {
-		glog.Exit("--log_url must not be empty")
-	}
-	der, err := base64.StdEncoding.DecodeString(*logPK)
-	if err != nil {
-		glog.Exitf("Failed to decode public key: %v", err)
-	}
-	pemPK := pem.EncodeToMemory(&pem.Block{
-		Type:  "PUBLIC KEY",
-		Bytes: der,
-	})
-	opts := jsonclient.Options{PublicKey: string(pemPK)}
-	c, err := client.New(*logURL, http.DefaultClient, opts)
-	if err != nil {
-		glog.Exitf("Failed to create JSON client: %v", err)
-	}
-	// Create the feeder with no initial witness STH.
+	// Create the feeder.
 	feeder := feeder{
-		logID: logID,
-		c:     c,
-		w:     w,
+		logs: logData,
+		w:    w,
 	}
-
+	// Now feed each log one by one.
 	tik := time.NewTicker(*interval)
 	for {
-		wSize := feeder.latestSize()
-		glog.V(2).Infof("Tick: start feedOnce (witness size %d)", wSize)
-		if err := feeder.feedOnce(ctx); err != nil {
-			glog.Warningf("Failed to feed: %v", err)
-		}
-		glog.V(2).Infof("Tick: feedOnce complete (witness size %d)", wSize)
+		for id, ld := range feeder.logs {
```

Comment:
This loop feels like it should be pushed into the feeder as a `feedAll` method. In general something that gets a bunch of Ys from object X, and then gives the Ys back to X indicates a poor abstraction (in this case, getting all of the IDs from the feeder, then calling a method on the feeder with each ID in turn).

Reply:
I'm not sure I fully see this, since the case switching at the end of this loop means it would be more of a `feedForever` method, that has to take the ticker as input, rather than just `feedAll` (but once).  Am I missing a way to neatly split apart the iteration through the logs and the switch?

Reply:
Thanks for flagging that. I'd actually misread this code and assumed the select was in the outer for loop. I think it would be better there, to the extent that it makes the poll interval easier to comprehend, and more consistent with the other tools. It's not the clearest to follow because there are a few layers of abstraction, but the serverless feeder (https://github.com/google/trillian-examples/blob/master/serverless/cmd/feeder/main.go#L72) does something similar to this and the flow here is:
  * `main` with outer loop iterating over all logs
  * Each loop starts a goroutine, X, for one of the logs
    * `main` only terminates when all Xs are done
  * Each X perpetually feeds its log at the polling interval

This separation of concerns is nice. It also keeps log feeding rate clear. In the approach implemented in this PR, the feeding interval for any log is `N * pollInterval`, where N is the total number of logs.

Reply:
Okay, I'm doing some version of this now!  Take a look and let me know what you think?